pinot-controller/src/main/java/com/linkedin/pinot/controller/ControllerConf.java

```diff
@@ -89,6 +102,24 @@ public static String constructDownloadUrl(String tableName, String segmentName,
     }
   }
 
+  public static String constructDownloadUrlWithVersion(String tableName, String segmentName, String vip, SegmentVersion segmentVersion) {
+    try {
+      return StringUtil.join("/", vip, "segments", tableName, URLEncoder.encode(segmentName, "UTF-8") + "?version=" + segmentVersion.toString());
+    } catch (UnsupportedEncodingException e) {
+      // Shouldn't happen
+      throw new AssertionError("UTF-8 encoding should always be supported", e);
```

Comment:
The error message doesn't read quite relevant. It might be best to say "Encountered error while encoding in UTF-8 format".